### DISCUSSION TITLE
Wasm import functions

### DISCUSSION THREAD

**Discussion by mikevoronov** [RustPython/RustPython#4866] [Q&A]
Sorry for possibly trivial Q, but I can't find any example or guide.

Is it possible to use WASI from Python code and provide some Wasm custom imports during VM initialization/running? Namely, can I call some GPU stuff from Python with f.e. `wasi-nn` or my own custom imports?

**Comment by mikevoronov** [comment]
Friendly reminder.

**Comment by mikevoronov** [comment]
So, it looks like a simple Q that developers should know the answer in advance. Or did I use an incorrect approach to reach you? If so, what is the best way?

**Comment by komodo472** [comment]
Yes its possible. You can define a `#[pyfunction]` like in this example 
https://github.com/RustPython/RustPython/blob/main/examples/call_between_rust_and_python.rs
where you call your own wasm custom imports. The src directory contains another example of how this this can be achieved. To build for wasi you just need to follow the instructions in the [README](https://github.com/RustPython/RustPython#wasi)
https://github.com/RustPython/RustPython/blob/f39f103dbec26cb2b8cdb803cb808d76a78786fa/src/lib.rs#L27-L32

**Reply by fanninpm** [reply]
@mikevoronov Did this answer your question?

**Comment by youknowone** [comment]
I don't know about `wasi-nn`, but if you can call it from rust side, you also can call it from RustPython as @komodo472 explained.
Thanks @komodo472 !
